File: frontend/src/types/ipc.ts
export type DataBits = 'seven' | 'eight';

export type Parity = 'none' | 'odd' | 'even';

export type StopBits = 'one' | 'two';

export type FlowControl = 'none' | 'software' | 'hardware';

export type ConnectionStatus = 'disconnected' | 'connecting' | 'connected' | 'error';

export interface IpcError {
  code: string;
  message: string;
  details?: string;
}

export type WriteError = 
  | { type: 'disconnected' }
  | { type: 'timeout' }
  | { type: 'io_error'; message: string };

export interface SerialParams {
  port: string;
  baud_rate: number;
  data_bits: DataBits;
  parity: Parity;
  stop_bits: StopBits;
  flow_control: FlowControl;
}

export interface TelnetParams {
  host: string;
  port: number;
  connect_timeout_secs: number;
}

export type ConnectionParams =
  | { type: 'serial' } & SerialParams
  | { type: 'telnet' } & TelnetParams;

export interface ConnectionStats {
  bytes_sent: number;
  bytes_received: number;
  packets_sent: number;
  packets_received: number;
}

export interface SessionInfo {
  id: string;
  name: string;
  connection_type: string;
  status: ConnectionStatus;
  created_at: string;
  last_activity?: string;
  stats: ConnectionStats;
  logging_enabled: boolean;
  log_file_path?: string;
}

export type LogDirection = 'input' | 'output';

export interface LoggingStatus {
  enabled: boolean;
  file_path?: string;
  bytes_logged_input: number;
  bytes_logged_output: number;
  started_at?: string;
}

export interface CommandInfo {
  index: number;
  name: string;
  description?: string;
  content_preview: string;
  line_number: number;
}

export interface SerialPortInfo {
  port_name: string;
  port_type: string;
  vendor_id?: number;
  product_id?: number;
  serial_number?: string;
  manufacturer?: string;
  product?: string;
}

export interface LogEntry {
  timestamp: number;
  session_id: string;
  direction: LogDirection;
  data: number[];
}

export interface CommandResult<T> {
  success: boolean;
  data?: T;
  error?: string;
}

export interface TabState {
  id: string;
  sessionId: string | null;
  title: string;
  isActive: boolean;
  isConnecting: boolean;
}

export interface TerminalUIState {
  showTimestamps: boolean;
  isSearchOpen: boolean;
  isConfigPanelOpen: boolean;
}

File: src-tauri/src/ipc.rs
//! Tauri IPC types for frontend-backend communication
//!
//!//! Re-exports lib types with IPC-specific additions.
//! Core types (SerialConfig, DataBits, etc.) are now directly used from lib.

use serde::{Deserialize, Serialize};

// Re-export core types from lib for convenience
pub use embedded_debugger::connection::types::{
    ConnectionStatus, SerialConfig, TelnetConfig, DataBits, Parity, StopBits, FlowControl,
};

/// IPC error type
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct IpcError {
    pub code: String,
    pub message: String,
    pub details: Option<String>,
}

impl IpcError {
    pub fn new(code: impl Into<String>, message: impl Into<String>) -> Self {
        Self {
            code: code.into(),
            message: message.into(),
            details: None,
        }
    }

    pub fn with_details(mut self, details: impl Into<String>) -> Self {
        self.details = Some(details.into());
        self
    }
}

/// IPC connection parameters (matches lib ConnectionConfig)
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ConnectionParams {
    Serial(SerialConfig),
    Telnet(TelnetConfig),
}

/// Serial port information for IPC
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SerialPortInfo {
    pub port_name: String,
    pub port_type: String,
    pub vendor_id: Option<u16>,
    pub product_id: Option<u16>,
    pub serial_number: Option<String>,
    pub manufacturer: Option<String>,
    pub product: Option<String>,
}

/// Connection statistics for IPC
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct ConnectionStats {
    pub bytes_sent: u64,
    pub bytes_received: u64,
    pub packets_sent: u64,
    pub packets_received: u64,
}

/// Log direction
#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum LogDirection {
    Input,
    Output,
}

/// Logging status
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LoggingStatus {
    pub enabled: bool,
    pub file_path: Option<String>,
    pub bytes_logged_input: u64,
    pub bytes_logged_output: u64,
    pub started_at: Option<String>,
}

/// Command information
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CommandInfo {
    pub index: usize,
    pub name: String,
    pub description: Option<String>,
    pub content_preview: String,
    pub line_number: usize,
}

/// Data received event payload
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DataReceivedEvent {
    pub connection_id: String,
    pub data: Vec<u8>,
}

/// Status changed event payload
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StatusChangedEvent {
    pub connection_id: String,
    pub status: ConnectionStatus,
}

/// Error occurred event payload
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ErrorOccurredEvent {
    pub connection_id: String,
    pub error: IpcError,
}
